### PR TITLE
Add new "Hero" component; includes variations of image alignment & art direction

### PR DESCRIPTION
**Pull request by sghoweri**
## Jira
http://vjira2:8080/browse/BDS-2016

## Summary
1st pass building several Hero variations / examples using off the shelf Bolt components + utility classes. 

## Details
Check out the `/pattern-lab/?p=viewall-components-hero` section in Pattern Lab to look at the full schema + demos.

> Note: updated per feedback from @mikemai2awesome and @danielamorse 

image

image

image

image

image

## How to test
- Review the props options + demos. Thoughts? Suggestions? Ideas?

### CODE REVIEW

File: docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/20-hero--reverse-order.twig [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        class: [
+          "u-bolt-line-length-medium"
+        ]
+      }
+    }),
+    macros.include("@bolt-components-button/button.twig", {
+      text: "See full agenda",
+      style: "secondary"
+    }),
+  ] | join(''),
```
Comment (danielamorse) [inline docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/20-hero--reverse-order.twig]: @sghoweri  not a blocker at all... I thought we decided to remove space around the pipe in Twig, per [Twig coding standards](https://twig.symfony.com/doc/2.x/coding_standards.html). That the convention I've been using (see [our wiki](https://github.com/boltdesignsystem/bolt/wiki/Twig%3A-Example-Component)).

File: packages/components/bolt-hero/hero.twig [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+          ]
+        }
+      } only %}
+    </div>
+  {% endset %}
+{% endif %}
+
+
+{% set content %}
+  {% grid "o-bolt-grid--flex o-bolt-grid--middle o-bolt-grid--medium #{   (reverseOrder == false and imageValign != 'bottom') or (reverseOrder == true and not image) ? 'o-bolt-grid--rev' : '' }" %}
```
Comment (danielamorse) [inline packages/components/bolt-hero/hero.twig]:
```suggestion
  {% grid "o-bolt-grid--flex o-bolt-grid--middle o-bolt-grid--medium #{(reverseOrder == false and imageValign != 'bottom') or (reverseOrder == true and not image) ? 'o-bolt-grid--rev' : '' }" %}
```

**Comment by sghoweri** [comment]
> @sghoweri Really nice job. These are looking great and the API makes sense to me.
> 
> A few comments (some overlap with @adamszalapski's):
> 
> * Why is Hero component prefixed with `10-`? That puts it at the top of the list. Can it be alphabetical like the rest?
> * I wonder if the props in "Advanced Schema Options" will be too easily missed. My preference would be to include them in the main Schema section at the bottom. Open to discussion.
> * Left a couple minor formatting comments. Not blockers, just if you have time.

Thanks @danielamorse -- I'm good with these suggestions 👍 

The number prefix for added earlier on to make it easy to pull this up -- I'll nix that when I move the "advanced schema options" out of the panel. 

Side thought: this might be something worth spending a little time on: if these advanced schema options aren't easily found (but we also don't want to clog up our docs pages) what are our options here... 🤔

File: docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/55-hero--image-valign-bottom.twig [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{% import "@bolt/macros.twig" as macros %}
+{# ^ allows us to use inline includes that also work in Drupal #}
+
+<bolt-text>This demo sets the <code>imageValign</code> prop to <code>bottom</code> so the image is always flush with the bottom-edge of the Hero container.</bolt-text>
+
+<aside class="c-bds-callout c-bds-callout--notice u-bolt-margin-bottom-large">
```
Comment (mikemai2awesome) [inline docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/55-hero--image-valign-bottom.twig]: Can we use `bolt-banner` for this?

File: packages/components/bolt-hero/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+A large promo section, generally at the top of a page, typically featuring a headline, call to action and a prominent image.
+
+<aside class="c-bds-callout c-bds-callout--notice u-bolt-margin-bottom-large">
```
Comment (mikemai2awesome) [inline packages/components/bolt-hero/README.md]: Also `bolt-banner` for this?

Comment (sghoweri) [reply packages/components/bolt-hero/README.md]: 👍🏼

File: docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/55-hero--image-valign-bottom.twig (hunk 3 above)
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/55-hero--image-valign-bottom.twig]: @mikemai2awesome sure, I just added it 👍🏼

File: docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/20-hero--reverse-order.twig (hunk 1 above)
Comment (sghoweri) [reply docs-site/src/pages/pattern-lab/_patterns/02-components/10-hero/20-hero--reverse-order.twig]: @danielamorse Okie dokie - just updated these

**Comment by sghoweri** [comment]
> @sghoweri Really nice job. These are looking great and the API makes sense to me.
> 
> A few comments (some overlap with @adamszalapski's):
> 
> * Why is Hero component prefixed with `10-`? That puts it at the top of the list. Can it be alphabetical like the rest?
> * I wonder if the props in "Advanced Schema Options" will be too easily missed. My preference would be to include them in the main Schema section at the bottom. Open to discussion.
> * Left a couple minor formatting comments. Not blockers, just if you have time.

@danielamorse I just pushed up some updates you should double check. These should take care of everything on your list.

**Comment by sghoweri** [comment]
> Good job @sghoweri i have only a few things as suggestions.
> 
> * In schema we don't have any information about `reverseOrder` and `imageMinWidth`

Check out the updated version of this @adamszalapski -- those props were previously there (tucked away in the Advanced Schema accordion) but should now be showing up normally.

> * I don't know if we should create a separate component for that maybe expand a band component

Happy to chat about this one further @adamszalapski but TLDR: Band is way way way too complicated + the internal Band updates these updates required would have caused a lot of internal damage to Band as a result. 

> * I was thinking about concatenating  content and image into one prop and passing data as array of elements like (image, content or content, image) that we will not need `reverseOrder` prop

I thought about this too however because Image requires very specific props set (ex. a different Grid cell with lots of conditional spacing, I worry that combining the Image + content data would either require even more configuration OR would be extremely fragile to use...

Definitely worth keeping an eye on + see if we can keep iterating in this approach.

> * I don't know why Hero is on the first place in the components menu and left menu

That was a temporary thing while I was doing the original build out of this -- code's been updated to now be in the correct order 🙂

**Comment by sghoweri** [comment]
PR was released with v2.19.0

**Pull request merged**